solidity/contracts/token/HypNative.sol

```diff
@@ -17,90 +32,120 @@ contract HypNative is TokenRouter {
      * @param amount The amount of native tokens donated.
      */
     event Donation(address indexed sender, uint256 amount);
+    // ============ Errors ============
+
+    error InsufficientValue(uint256 requiredValue, uint256 providedValue);
 
     constructor(address _mailbox) TokenRouter(_mailbox) {}
 
+    // ============ Initialization ============
+
     /**
-     * @notice Initializes the Hyperlane router
-     * @param _hook The post-dispatch hook contract.
-       @param _interchainSecurityModule The interchain security module contract.
-       @param _owner The this contract.
+     * @notice Initializes the contract
+     * @param _valuehook The address of the value transfer hook
+     * @param _interchainSecurityModule The address of the interchain security module
+     * @param _owner The owner of the contract
      */
     function initialize(
-        address _hook,
+        address _valuehook,
         address _interchainSecurityModule,
         address _owner
     ) public initializer {
-        _MailboxClient_initialize(_hook, _interchainSecurityModule, _owner);
+        _MailboxClient_initialize(
+            _valuehook,
+            _interchainSecurityModule,
+            _owner
+        );
     }
 
-    /**
-     * @inheritdoc TokenRouter
-     * @dev uses (`msg.value` - `_amount`) as hook payment and `msg.sender` as refund address.
-     */
-    function transferRemote(
-        uint32 _destination,
-        bytes32 _recipient,
-        uint256 _amount
-    ) external payable virtual override returns (bytes32 messageId) {
-        require(msg.value >= _amount, "Native: amount exceeds msg.value");
-        uint256 _hookPayment = msg.value - _amount;
-        return _transferRemote(_destination, _recipient, _amount, _hookPayment);
-    }
+    // ============ External Functions ============
 
     /**
      * @inheritdoc TokenRouter
-     * @dev uses (`msg.value` - `_amount`) as hook payment.
+     * @dev use _hook with caution, make sure that this hook can handle msg.value transfer using the metadata.msgValue()
      */
     function transferRemote(
         uint32 _destination,
         bytes32 _recipient,
         uint256 _amount,
         bytes calldata _hookMetadata,
         address _hook
-    ) external payable virtual override returns (bytes32 messageId) {
-        require(msg.value >= _amount, "Native: amount exceeds msg.value");
-        uint256 _hookPayment = msg.value - _amount;
+    ) public payable virtual override returns (bytes32 messageId) {
+        uint256 quote = _GasRouter_quoteDispatch(
+            _destination,
+            _hookMetadata,
+            _hook
+        );
+        require(msg.value >= _amount + quote, "HypNative: insufficient value");
```

Comment:
why did you change the conditions here?
previously we deferred to the hook to enforce the quote and I think that is desirable

Reply:
done